Gems/Atom/RPI/Code/Source/RPI.Reflect/Material/MaterialAsset.cpp

```diff
@@ -102,32 +103,114 @@ namespace AZ
+
+                    if (value.Is<AZStd::string>() && propertyDescriptor->GetDataType() == MaterialPropertyDataType::Enum)
+                    {
+                        AZ::Name enumName = AZ::Name(value.GetValue<AZStd::string>());
+                        uint32_t enumValue = propertyDescriptor->GetEnumValue(enumName);
+                        if (enumValue == MaterialPropertyDescriptor::InvalidEnumValue)
+                        {
+                            reportWarning(AZStd::string::format("Material property name \"%s\" has invalid enum value \"%s\".", name.GetCStr(), enumName.GetCStr()).c_str());
+                        }
+                        else
+                        {
+                            finalizedPropertyValues[propertyIndex.GetIndex()] = enumValue;
+                        }
+                    }
+                    else if (value.Is<AZStd::string>() && propertyDescriptor->GetDataType() == MaterialPropertyDataType::Image)
+                    {
+                        // Here we assume that the material asset builder resolved any image source file paths to an ImageAsset reference.
+                        // So the only way a string could be present is if it's an empty image path reference, meaning no image should be bound.
+                        AZ_Assert(value.GetValue<AZStd::string>().empty(), "Material property '%s' references in image '%s'. Image file paths must be resolved by the material asset builder.");
 
-                if (m_isDirty)
+                        finalizedPropertyValues[propertyIndex.GetIndex()] = Data::Asset<ImageAsset>{};
+                    }
+                    else
+                    {
+                        if (ValidateMaterialPropertyDataType(value.GetTypeId(), name, propertyDescriptor, reportError))
+                        {
+                            finalizedPropertyValues[propertyIndex.GetIndex()] = value;
+                        }
+                    }
+                }
+                else
                 {
-                    const_cast<MaterialAsset*>(this)->RealignPropertyValuesAndNames();
+                    reportWarning(AZStd::string::format("Material property name \"%s\" is not found in the material properties layout and will not be used.", name.GetCStr()).c_str());
                 }
             }
 
+            m_propertyValues.swap(finalizedPropertyValues);
+
+            m_isFinalized = true;
+        }
+
+        const AZStd::vector<MaterialPropertyValue>& MaterialAsset::GetPropertyValues() const
+        {
+            // This can't be done in MaterialAssetHandler::LoadAssetData because the MaterialTypeAsset isn't necessarily loaded at that point.
+            // And it can't be done in PostLoadInit() because that happens on the next frame which might be too late. So we finalize just-in-time
+            // when properties are accessed.
+            const_cast<MaterialAsset*>(this)->Finalize();
```

Comment:
Note: I believe this may be invoking undefined behavior.

From [cppreference](https://en.cppreference.com/w/cpp/language/const_cast)
> Modifying a const object through a non-const access path and referring to a volatile object through a non-volatile glvalue results in undefined behavior.

I would suggest just making this method non-const or adding the correct `mutable` decorators as needed

Reply:
As I understand it, it would only be unsafe if the original MaterialAsset object was const, which I think is never the case. Even then, I'm not sure I understand all the nuances here. This has some interesting discussion https://stackoverflow.com/questions/357600/is-const-cast-safe . Anyway, I'll move things around to avoid doing this, just in case, thanks for pointing it out.

FWIW, the most common place where I'm used to doing const_cast is in const/non-const pairs of getter functions where you want to factor out code, like "Item* Find(ID)" and "const Item* Find(ID) const". But in that case, I guess you are doing the const_cast inside the non-const function, thus ensuring the object is indeed non-const and clearly avoiding any undefined behavior.